streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.util
+
+import java.nio.ByteBuffer
+import java.util.concurrent.{LinkedBlockingQueue, TimeoutException}
+import java.util.{Iterator => JIterator}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{Await, ExecutionContext, Promise}
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+import org.apache.spark.Logging
+import org.apache.spark.util.{Utils, ThreadUtils}
+
+/**
+ * A wrapper for a WriteAheadLog that batches records before writing data. All other methods will
+ * be passed on to the wrapped class.
+ *
+ * Parent exposed for testing.
+ */
+private[streaming] class BatchedWriteAheadLog(private[util] val parent: WriteAheadLog)
+  extends WriteAheadLog with Logging {
+
+  import BatchedWriteAheadLog._
+
+  // exposed for tests
+  protected val walWriteQueue = new LinkedBlockingQueue[RecordBuffer]()
+
+  private val WAL_WRITE_STATUS_TIMEOUT = 5000 // 5 seconds
+
+  // Whether the writer thread is active
+  @volatile private var active: Boolean = true
+  protected val buffer = new ArrayBuffer[RecordBuffer]()
```

Comment:
It has a single asJava call. We use the beauty of methods like Seq.map, tail, everywhere else. I think this should stay as an ArrayBuffer. Too much looping overhead otherwise